File: source/uems-agent-chat/src/tools/search-repos.ts
/**
 * Tool: uems_search_repos
 * Search for code patterns across multiple workspace repos simultaneously.
 */

import * as vscode from 'vscode';
import { searchRepos, normalizeQuery } from '../core/search-engine';
import { resolveRepoPaths, jsonResult, truncate } from './helpers';
import { lookupRepoPlatform } from '../core/repo-registry';

interface SearchReposInput {
  query: string;
  repos?: string[];
  filePattern?: string;
  maxResults?: number;
  isRegex?: boolean;
}

export class SearchReposTool implements vscode.LanguageModelTool<SearchReposInput> {

  async invoke(
    options: vscode.LanguageModelToolInvocationOptions<SearchReposInput>,
    token: vscode.CancellationToken,
  ): Promise<vscode.LanguageModelToolResult> {
    void token;
    const { repos, filePattern, maxResults } = options.input;
    const repoPaths = resolveRepoPaths(repos);

    if (repoPaths.length === 0) {
      return jsonResult({ error: 'No matching repos found in workspace.' });
    }

    // Normalize multi-term queries: space, |, & separated terms → regex alternation
    const normalized = normalizeQuery(options.input.query, options.input.isRegex ?? false);

    const result = await searchRepos({
      query: normalized.query,
      repoPaths,
      filePattern,
      maxResults: maxResults ?? 200,
      isRegex: normalized.isRegex,
    });

    // Group by repo → compact "file:line:text" strings
    const grouped: Record<string, string[]> = {};
    for (const m of result.matches) {
      if (!grouped[m.repo]) { grouped[m.repo] = []; }
      grouped[m.repo].push(`${m.file}:${m.line}:${truncate(m.text.trim())}`);
    }

    // Attach platform metadata for each repo so the agent knows which platform
    // a .h / .swift / .go file belongs to without relying on file extension alone.
    const repoPlatforms: Record<string, string> = {};
    for (const repoName of Object.keys(grouped)) {
      const platform = lookupRepoPlatform(repoName);
      if (platform) { repoPlatforms[repoName] = platform; }
    }

    return jsonResult({
      total: result.matches.length,
      truncated: result.truncated,
      ...(Object.keys(repoPlatforms).length > 0 && { repoPlatforms }),
      results: grouped,
    });
  }
}

File: source/uems-agent-web/backend/tools/search-repos.go
package tools

import (
	"context"
	"fmt"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type searchReposInput struct {
	Query       string   `json:"query" jsonschema:"Search pattern. Multiple terms can be separated by spaces, pipes (|), or ampersands (&) to match any of them (OR logic). Examples: 'WriteEvent ERRMSG' finds both, 'NSXPCConnection' finds exact match, 'init.*Logger' with isRegex=true for advanced regex."`
	Repos       []string `json:"repos,omitempty" jsonschema:"Repo names to search (e.g. ['agent-utils', 'patch-management']). Searches all workspace repos if omitted."`
	FilePattern string   `json:"filePattern,omitempty" jsonschema:"Glob pattern for files (e.g. '*.swift', '*.{h,cpp}')."`
	MaxResults  int      `json:"maxResults,omitempty" jsonschema:"Max results to return (default 200)."`
	IsRegex     bool     `json:"isRegex,omitempty" jsonschema:"Set to true only for advanced regex syntax (character classes, quantifiers, lookaheads). NOT needed for multi-term searches. Default: false."`
}

func registerSearchRepos(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name:        "uems_agent_search_repos",
		Description: "Search for code patterns across UEMS agent repositories. Supports multiple search terms separated by spaces, pipes (|), or ampersands (&) — each term is matched independently (OR logic). Set isRegex=true only for advanced regex patterns.",
	}, func(_ context.Context, _ *mcp.CallToolRequest, input searchReposInput) (*mcp.CallToolResult, any, error) {
		repos := input.Repos
		if len(repos) == 0 {
			repos = allRepoNames()
		}
		repoPaths := resolveRepoPaths(repos)
		if len(repoPaths) == 0 {
			return errorResult("No matching repos found on disk. Ensure repos are cloned."), nil, nil
		}

		maxR := input.MaxResults
		if maxR <= 0 {
			maxR = 200
		}

		// Normalize multi-term queries: space, &, | separated terms → regex alternation
		query, isRegex := normalizeQuery(input.Query, input.IsRegex)

		matches, truncated := rgSearch(query, repoPaths, input.FilePattern, maxR, isRegex)

		grouped := map[string][]string{}
		for _, m := range matches {
			grouped[m.Repo] = append(grouped[m.Repo], fmt.Sprintf("%s:%d:%s", m.File, m.Line, m.Text))
		}

		// Attach platform metadata for each repo
		repoPlatforms := map[string]string{}
		for repoName := range grouped {
			for platform, platformRepos := range registry {
				if _, ok := platformRepos[repoName]; ok {
					repoPlatforms[repoName] = platform
					break
				}
			}
		}

		resp := map[string]interface{}{
			"total":     len(matches),
			"truncated": truncated,
			"results":   grouped,
		}
		if len(repoPlatforms) > 0 {
			resp["repoPlatforms"] = repoPlatforms
		}

		return jsonResponse(resp), nil, nil
	})
}
